src/main/java/org/broadinstitute/hellbender/tools/spark/RevertSamSpark.java

```diff
@@ -0,0 +1,706 @@
+                                                  inHeader.getReadGroups(),
+                                                  outputByReadGroup);
+
+        ////////////////////////////////////////////////////////////////////////////
+        // Construct appropriate headers for the output files
+        ////////////////////////////////////////////////////////////////////////////
+        final Map<String, SAMFileHeader> headerMap = getReadgroupHeaderMap(inHeader, writerMap);
+
+        // Revert the reads based on the given attributes
+        List<String> attributesToRevert = removeDefaults? DEFAULT_ATTRIBUTES_TO_CLEAR: new ArrayList<>();
+        attributesToRevert.addAll(attributesToClear);
+        JavaRDD<GATKRead> readsReverted = revertReads(reads, attributesToRevert);
+
+        ////////////////////////////////////////////////////////////////////////////
+        // Sanitize the reads, sorting them into appropriate order if necessary
+        ////////////////////////////////////////////////////////////////////////////
+        if (sanitize) {
+            Map<String, FastqQualityFormat> readGroupFormatMap = createReadGroupFormatMap(readsReverted, headerb, !dontRestoreOriginalQualities);
+
+            readsReverted = sanitize(readGroupFormatMap, readsReverted, inHeader, keepFirstDuplicate);
+        }
+
+        // Write the one or many read output files
+        for (Map.Entry<String, File> rmap: writerMap.entrySet()) {
+            //TODO what to do if the readgroup isn't present
+            final String key = rmap.getKey();
+            JavaRDD<GATKRead> filteredreads = rmap.getKey()==null? readsReverted :
+                                                                    readsReverted.filter(r -> r.getReadGroup().equals(key));
+            writeReads(ctx, rmap.getValue().getPath(), filteredreads, headerMap.get(rmap.getKey())); //TODO proper header map
+        }
+    }
+
+    /**
+     * Runs the QualityEncodingDetector over a sampling of each readGroup present in the file to detect what the encoding format
+     * for base quality those reads are.
+     *
+     * @param reads Reads RDD over which to iterate and detect readgroups
+     * @param inHeader Header describing the readgroups present in the bam
+     * @param restoreOriginalQualities Whether to use the OQ tag for determining the map
+     * @return
+     */
+    private Map<String, FastqQualityFormat> createReadGroupFormatMap( final JavaRDD<GATKRead> reads,
+                                                                      final Broadcast<SAMFileHeader> inHeader,
+                                                                      final boolean restoreOriginalQualities) {
+        final Map<String, FastqQualityFormat> output = new HashMap<>();
+
+        inHeader.getValue().getReadGroups().stream().parallel().forEach(rg -> {
```

Comment:
It's kind of weird to do spark operations in a parallel stream.  Is this safe?  Is it faster?  Spark already tries to use all the resources, it seems like this would just create more contention.

Reply:
Is spark threadsafe?